wpimath/src/main/native/include/frc/MathUtil.h

```diff
@@ -94,24 +94,43 @@ constexpr T ApplyDeadband(T value, T deadband, T maxMagnitude = T{1.0}) {
   }
 }
 
+/**
+ * Returns a zero vector if the given vector is within the specified
+ * distance from the origin. The remaining distance between the deadband and the
+ * maximum distance is scaled from the origin to the maximum distance.
+ *
+ * @param value Value to clip.
+ * @param deadband Distance from origin.
+ * @param maxMagnitude The maximum distance from the origin of the input
+ * (defaults to 1). Can be infinite.
+ * @return The value after the deadband is applied.
+ */
+template <int N>
+Eigen::Matrix<double, N, 1> ApplyDeadband(
+    const Eigen::Matrix<double, N, 1>& value, double deadband,
+    double maxMagnitude = 1.0) {
+  if (value.norm() < 1e-6) {
+    return Eigen::Matrix<double, N, 1>::Zero();
+  }
+  return value.normalized() *
+         ApplyDeadband(value.norm(), deadband, maxMagnitude);
+}
```

Comment:
The generic version should be something like this:
```suggestion
template <typename T, int N>
Eigen::Vector<T, N> ApplyDeadband(
    const Eigen::Vector<T, N>& value, T deadband,
    T maxMagnitude = T{1.0}) {
  if (value.norm() < T{1e-6}) {
    return Eigen::Vector<T, N>::Zero();
  }
  return value.normalized() *
         ApplyDeadband(value.norm(), deadband, maxMagnitude);
}
```

Reply:
I had something like that [here (https://github.com/wpilibsuite/allwpilib/pull/8057/commits/cbf57fef19d81cc92cbcc509010aea721049855a), but would get errors that C++ was unsure of what T was, because the non vector version also accepts a templated first param

Reply:
trying it now though, will come back and add some unit tests in a bit

Reply:
I think that Eigen does not know how to take the norm (sqrt) with units

compiling MathUtilTest.cpp failed.
MathUtilTest.cpp
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/MathFunctions.h(288): error C2665: 'sqrt': no overloaded function could convert all the argument types
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\include\cmath(530): note: could be 'long double sqrt(long double) noexcept'
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/MathFunctions.h(288): note: 'long double sqrt(long double) noexcept': cannot convert argument 1 from 'const Scalar' to 'long double'
        with
        [
            Scalar=units::unit_t<units::velocity::meters_per_second,double,units::linear_scale>
        ]
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/MathFunctions.h(288): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\include\cmath(264): note: or       'float sqrt(float) noexcept'
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/MathFunctions.h(288): note: 'float sqrt(float) noexcept': cannot convert argument 1 from 'const Scalar' to 'float'
        with
        [
            Scalar=units::unit_t<units::velocity::meters_per_second,double,units::linear_scale>
        ]
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/MathFunctions.h(288): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
C:\Program Files (x86)\Windows Kits\10\Include\10.0.26100.0\ucrt\corecrt_math.h(519): note: or       'double sqrt(double)'
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/MathFunctions.h(288): note: 'double sqrt(double)': cannot convert argument 1 from 'const Scalar' to 'double'
        with
        [
            Scalar=units::unit_t<units::velocity::meters_per_second,double,units::linear_scale>
        ]
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/MathFunctions.h(288): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\include\complex(1577): note: or       'std::complex<_Ty> std::sqrt(const std::complex<_Ty> &) noexcept(<expr>)'
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/MathFunctions.h(288): note: 'std::complex<_Ty> std::sqrt(const std::complex<_Ty> &) noexcept(<expr>)': could not deduce template argument for 'const std::complex<_Ty> &' from 'const Scalar'
        with
        [
            Scalar=units::unit_t<units::velocity::meters_per_second,double,units::linear_scale>
        ]
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\include\cmath(866): note: or       'double sqrt(_Ty) noexcept'
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/MathFunctions.h(288): note: 'double sqrt(_Ty) noexcept': could not deduce template argument for '__formal'
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\include\cstddef(36): note: 'std::enable_if_t<false,int>' : Failed to specialize alias template
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/MathFunctions.h(288): note: while trying to match the argument list '(const Scalar)'
        with
        [
            Scalar=units::unit_t<units::velocity::meters_per_second,double,units::linear_scale>
        ]
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/MathFunctions.h(288): note: the template instantiation context (the oldest one first) is
~\allwpilib\wpimath\src\test\native\cpp\MathUtilTest.cpp(119): note: see reference to function template instantiation 'Eigen::Matrix<units::velocity::meters_per_second_t,2,1,0,2,1> frc::ApplyDeadband<units::velocity::meters_per_second_t,2>(const Eigen::Matrix<units::velocity::meters_per_second_t,2,1,0,2,1> &,T,T)' being compiled
        with
        [
            T=units::velocity::meters_per_second_t
        ]
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/Dot.h(78): note: while compiling class template member function 'units::unit_t<units::velocity::meters_per_second,double,units::linear_scale> Eigen::MatrixBase<Derived>::norm(void) const'
        with
        [
            Derived=Eigen::Matrix<units::velocity::meters_per_second_t,2,1,0,2,1>
        ]
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/Dot.h(79): note: see reference to function template instantiation 'units::unit_t<units::velocity::meters_per_second,double,units::linear_scale> Eigen::numext::sqrt<units::unit_t<units::velocity::meters_per_second,double,units::linear_scale>>(const Scalar &)' being compiled
        with
        [
            Scalar=units::unit_t<units::velocity::meters_per_second,double,units::linear_scale>
        ]
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/MathFunctions.h(1349): note: see reference to class template instantiation 'Eigen::internal::sqrt_impl<units::unit_t<units::velocity::meters_per_second,double,units::linear_scale>>' being compiled
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/MathFunctions.h(286): note: while compiling class template member function 'Scalar Eigen::internal::sqrt_impl<Scalar>::run(const Scalar &)'
        with
        [
            Scalar=units::unit_t<units::velocity::meters_per_second,double,units::linear_scale>
        ]
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/MathFunctions.h(1349): note: see the first reference to 'Eigen::internal::sqrt_impl<units::unit_t<units::velocity::meters_per_second,double,units::linear_scale>>::run' in 'Eigen::numext::sqrt'
~\allwpilib\wpimath\src\main\native\thirdparty\eigen\include\Eigen\src/Core/Dot.h(79): note: see the first reference to 'Eigen::numext::sqrt' in 'Eigen::MatrixBase<Derived>::norm'
        with
        [
            Derived=Eigen::Matrix<units::velocity::meters_per_second_t,2,1,0,2,1>
        ]
~\allwpilib\wpimath\src\main\native\include\frc/MathUtil.h(112): note: see the first reference to 'Eigen::MatrixBase<Derived>::norm' in 'frc::ApplyDeadband'
        with
        [
            Derived=Eigen::Matrix<units::velocity::meters_per_second_t,2,1,0,2,1>
        ]

Reply:
I believe that worked, but the deeper issue I just noticed is that doing math with units does not always return the same unit (I think, based on looking though https://github.com/wpilibsuite/allwpilib/actions/runs/17204129980/job/48800895396?pr=8057#step:4:440)